### PR TITLE
Nll loss converter

### PR DESCRIPTION
**Pull request by bpickrel**
Adds Migraphx converters for the NLL Loss function; of its many implementations, we use those in `torch.ops.aten.nll_loss_forward.default` and in `torch.nn.functional.nll_loss`

My converter introduces a Python `GatherElements()` function which we can use in future implementation of the Onnx op.

Closes Op Support: aten.nll_loss_forward.default
[#94](https://github.com/ROCm/torch_migraphx/issues/94)

### CODE REVIEW

File: py/torch_migraphx/fx/converters/acc_ops_converters.py [hunk 1]
```diff
@@ -75,6 +225,19 @@ def broadcast_for_elemwise_op(mgx_module,
 
     return inp, other
 
+def broadcast_scalar_for_elemwise_op(mgx_module,
```
Comment (bpickrel) [inline py/torch_migraphx/fx/converters/acc_ops_converters.py:228]: not used

File: py/torch_migraphx/fx/converters/acc_ops_converters.py [hunk 2]
```diff
@@ -33,14 +33,164 @@
...
+							  
+# 							                                Output[x1, x2, ... xA, ... x(ndim-1)] = ?
+															
+# 							    where xA is the coordinate on the "axis" dimension; substitute the value from "index":
+							  
+# 							                                A_ind = index[x1, x2, ... xA, ... x(ndim-1)] 
+# 							    then fetch 
+#                                                             data[x1, x2, ... A_ind, ... x(ndim-1)]							  
+
+def gather_elements(info, axis, args):
```
Comment (shivadbhavsar) [inline py/torch_migraphx/fx/converters/acc_ops_converters.py:139]: Can we move this and all the helper functions to the `utils.py` file in this directory. Thats where the current helper functions reside.

File: py/torch_migraphx/fx/tracer/acc_tracer/acc_ops.py [hunk 3]
```diff
@@ -261,6 +261,26 @@ def select_scatter(*, input, src, dim, index):
...
+        ("size_average", "size_average", this_arg_is_optional),
+        ("reduce", "reduce", this_arg_is_optional),
+        ("reduction", "reduction", this_arg_is_optional),
+        ("ignore_index" , "ignore_index" , this_arg_is_optional)
+    ],
+)
+# defines the list of arguments recognized when we define a forward 
+# function in a torch.nn.Module .
+@register_acc_op
+def nll_loss_forward(*, input, target, weight=None, reduce=None, reduction='mean', size_average=None, ignore_index=-100):
```
Comment (shivadbhavsar) [inline py/torch_migraphx/fx/tracer/acc_tracer/acc_ops.py:279]:
```suggestion
def nll_loss(*, input, target, weight=None, reduce=None, reduction='mean', size_average=None, ignore_index=-100):
```

Just to keep the naming of these functions consistent

File: py/torch_migraphx/fx/converters/aten_ops_converters.py [hunk 4]
```diff
@@ -296,6 +296,42 @@ def aten_ops_clamp(mgx_module, node, args, _kwargs):
     return acc_ops_converters.acc_ops_clamp(mgx_module, node, (), acc_kwargs)
 
 
+@migraphx_converter(torch.ops.aten.nll_loss_forward.default)
+def aten_ops_nllloss(mgx_module, node, args, _kwargs):
```
Comment (shivadbhavsar) [inline py/torch_migraphx/fx/converters/aten_ops_converters.py:300]:
```suggestion
def aten_ops_nll_loss_forward(mgx_module, node, args, _kwargs):
```

Naming consistency

File: py/torch_migraphx/fx/converters/acc_ops_converters.py [hunk 5]
```diff
@@ -109,6 +259,74 @@ def acc_ops_linear(mgx_module, node, args, kwargs):
     return MGXInstruction(out_mgx)
 
 
+# NLL (negative log likelihood loss) converter.  This op assumes data has already been normalized with log_softmax
+# to give meaningful results, although there is no restriction on input values.
+@migraphx_converter(acc_ops.nll_loss_forward)
+def acc_ops_nll_loss_forward(mgx_module, node, args, kwargs):
```
Comment (shivadbhavsar) [inline py/torch_migraphx/fx/converters/acc_ops_converters.py:265]:
```suggestion
@migraphx_converter(acc_ops.nll_loss)
def acc_ops_nll_loss(mgx_module, node, args, kwargs):
```
Necessary based on other name change requested

File: tests/fx/converters/test_activations_fx.py [hunk 6]
```diff
@@ -3,6 +3,42 @@
...
+    mod1 = FuncModule(torch.nn.functional.nll_loss, target=target, weight=weight,
+                       reduction = 'mean', ignore_index = -100)
+    mod2 = FuncModule(torch.nn.functional.nll_loss, target=target, weight=weight,
+                       reduction = 'sum', ignore_index = -100)
+    mod3 = FuncModule(torch.nn.functional.nll_loss, target=target, weight=weight,
+                       reduction = 'none', ignore_index = -100)
+
+    for mod in [mod1, mod2, mod3]:
+        mgx_mod = convert_to_mgx(mod, [inp])
+        verify_outputs(mod, mgx_mod, [inp])
```
Comment (shivadbhavsar) [inline tests/fx/converters/test_activations_fx.py:39]:
```suggestion
    mod = FuncModule(torch.nn.functional.nll_loss, target=target, weight=weight,
                       reduction=reduction, ignore_index = -100)

    mgx_mod = convert_to_mgx(mod, [inp])
    verify_outputs(mod, mgx_mod, [inp])
```

The reduction parameter here should be parametrized. Right now even if one of the reduction type fails, the whole test will fail and you will have no idea which one of these modules failed.

File: py/torch_migraphx/fx/converters/acc_ops_converters.py [hunk 7]
```diff
@@ -33,14 +33,164 @@
...
+							  
+# 							                                A_ind = index[x1, x2, ... xA, ... x(ndim-1)] 
+# 							    then fetch 
+#                                                             data[x1, x2, ... A_ind, ... x(ndim-1)]							  
+
+def gather_elements(info, axis, args):
+    arg_data = args[0]
+    arg_ind = args[1]
+
+    # todo: how to deal with nonstandard shapes?  The C++ version inserts two make_contiguous()
```
Comment (shivadbhavsar) [inline py/torch_migraphx/fx/converters/acc_ops_converters.py:143]: I think we can insert the contiguous calls in the same places as the c++ code. We have a pass that eliminates unnecessary contiguous calls anyways. If we dont, we might run into issues if this is called after a transpose op maybe

File: py/torch_migraphx/fx/converters/acc_ops_converters.py [hunk 8]
```diff
@@ -112,6 +113,74 @@ def acc_ops_linear(mgx_module, node, args, kwargs):
...
+            migraphx.op('unsqueeze', axes=list(range(ndims-1))), [weight])
+    weight_bcst = mgx_module.add_instruction(
+            migraphx.op('multibroadcast', out_lens=inp_instr_ref.shape().lens()), [weight_unsquoze])
+
+    # Use target indexes to select weights, one for each class
+    class_axis = 1
+    weight_to_use = gather_elements(mgx_module, class_axis, [weight_bcst, target_ref_unsquoze])
+
+    # Use target indexes to select inputs
+    x_to_use = gather_elements(mgx_module, class_axis, [neg_ins, target_ref_unsquoze])
```
Comment (shivadbhavsar) [inline py/torch_migraphx/fx/converters/acc_ops_converters.py:151]:
```suggestion
    target_unsq = mgx_module.add_instruction(migraphx.op('unsqueeze', axes=[-1]), [target_ref])
    batch_dim_indices = mgx_module.add_literal(torch.arange(target_ref.shape().elements()).numpy())
    batch_dim_indices_unsq = mgx_module.add_instruction(migraphx.op('unsqueeze', axes=[-1]), [batch_dim_indices])
    gathernd_indices = mgx_module.add_instruction(migraphx.op('concat', axis=-1), [batch_dim_indices_unsq, target_unsq])
    
    weight_to_use = mgx_module.add_instruction(migraphx.op('gathernd'), [weight_bcst, gathernd_indices])
    x_to_use = mgx_module.add_instruction(migraphx.op('gathernd'), [neg_ins, gathernd_indices])
```
You can use the `gathernd` op here instead of the complicated stuff going on in the util functions. This should work, you'll just have to fix the reduction axis on the reduction ops below.

File: py/torch_migraphx/fx/converters/acc_ops_converters.py [hunk 9]
```diff
@@ -112,6 +113,74 @@ def acc_ops_linear(mgx_module, node, args, kwargs):
...
+        # Don't reduce; return a 1-d vector
+        squeezed_multiply_ins =  mgx_module.add_instruction(migraphx.op('squeeze'), [multiply_ins])
+        return MGXInstruction(squeezed_multiply_ins)
+    else:
+        #
+        #    sum- or mean-reduction case.  Sum W * X, divide by sum of weights, and return a scalar
+        #
+
+        # Reduce, i.e. take the sum of all values 
+        reduce_ins =  mgx_module.add_instruction(migraphx.op('reduce_sum', axes=list(range(ndims))), [multiply_ins])
```
Comment (shivadbhavsar) [inline py/torch_migraphx/fx/converters/acc_ops_converters.py:167]: for axis just use something like `list(range(multiply_ins.shape().ndim()))`. That way you are not relying on the shape from gather op to be rank ndims

File: py/torch_migraphx/fx/converters/acc_ops_converters.py [hunk 10]
```diff
@@ -112,6 +113,74 @@ def acc_ops_linear(mgx_module, node, args, kwargs):
...
+        # squeeze the number of dimensions down to none (i.e. scalar)
+        squeezed_reduce_ins =  mgx_module.add_instruction(migraphx.op('squeeze'), [reduce_ins])
+        if kwargs.get('reduction') == 'sum':
+            return MGXInstruction(squeezed_reduce_ins)
+        #
+        # the default reduction type is 'mean'
+        #
+        # Calculate the sum of weights.  weight_to_use is a 1-d vector
+        sum_ins = mgx_module.add_instruction(migraphx.op('reduce_sum', axes=[0]), [weight_to_use])
```
Comment (shivadbhavsar) [inline py/torch_migraphx/fx/converters/acc_ops_converters.py:176]: With the gathernd change, you can just do `list(range(weight_to_use.shape().ndim()))` for the axis as it might not be 1 dim.

**Review by shivadbhavsar — DISMISSED** [review]

**Comment by shivadbhavsar** [comment]
Marked comments related to `utils.py` as resolved as it wont be necessary if the suggested gathernd approach is used

File: py/torch_migraphx/fx/converters/acc_ops_converters.py (hunk 8 above)
Comment (bpickrel) [reply py/torch_migraphx/fx/converters/acc_ops_converters.py:151]: I need to understand how this works better to know if it's generalizable to k dims.   It looks like it's specific to certain ranks of inputs, 1 or 2 or both?

I believe the Migraphx `unsqueeze` op can support more than one input axis, but only one negative axis input.  Not sure if that would come into play here.

Comment (shivadbhavsar) [reply py/torch_migraphx/fx/converters/acc_ops_converters.py:151]: actually, this might work as it is for k dims. The `gathernd_indices` shouldnt need to change for the k dims case (thats what the unsqueeze is used for)

Comment (shivadbhavsar) [reply py/torch_migraphx/fx/converters/acc_ops_converters.py:151]: only thing that might need to be modified is the output shape in the case of no reduction (although I think that should also be correct as written)

Comment (bpickrel) [reply py/torch_migraphx/fx/converters/acc_ops_converters.py:151]: > actually, this might work as it is for k dims. The `gathernd_indices` shouldnt need to change for the k dims case (thats what the unsqueeze is used for)

a little testing shows that `batch_dim_indices` above doesn't work as a 1-d index and would have to be created from the dimension 0 range instead, and then reshaped with more dimensions added in a way that GatherND could read the result correctly.  I think it could eventually be made to work but it's starting to look nearly as complicated as the `GatherElements` version.  So this change doesn't save us much.

Comment (shivadbhavsar) [reply py/torch_migraphx/fx/converters/acc_ops_converters.py:151]: Well taking another look, it doesnt just work as it is because the target tensor is also extended by kdims. You can generalize it by treating all the N + k-dims as batch dims (apply transpose to the input so that C is the last dim) and creating a k-dimensional base coordinate grid for it rather than the 1-d I showed originally.

Comment (shivadbhavsar) [reply py/torch_migraphx/fx/converters/acc_ops_converters.py:151]: I messed around a bit in my local environment, and this seems to work (might need ironing out but this is the concept):

```
target_unsq = mgx_module.add_instruction(migraphx.op('unsqueeze', axes=[-1]), [target_ref])
    target_rank = target_unsq.shape().ndim()
    # Create base coordinate tensors
    coordinate_index_lits = []
    for ax in range(target_rank-1):
        batch_dim_indices = mgx_module.add_literal(torch.arange(target_unsq.shape().lens()[ax]).numpy())
        unsq_dims = list(range(target_rank))
        unsq_dims = unsq_dims[:ax] + unsq_dims[ax+1:]
        
        batch_dim_indices_unsq = mgx_module.add_instruction(migraphx.op('unsqueeze', axes=unsq_dims), [batch_dim_indices])
        batch_dim_indices_bc = mgx_module.add_instruction(
            migraphx.op('multibroadcast', out_lens=target_unsq.shape().lens()), [batch_dim_indices_unsq])
        coordinate_index_lits.append(batch_dim_indices_bc)

    # Add the target indices to the base coordinates
    coordinate_index_lits.append(target_unsq)
    gathernd_indices = mgx_module.add_instruction(migraphx.op('concat', axis=-1), coordinate_index_lits)
    
    perm = [0] + list(range(2, ndims)) + [1]
    weight_bcst_t = mgx_module.add_instruction(migraphx.op('transpose', permutation=perm), [weight_bcst])
    neg_ins_t = mgx_module.add_instruction(migraphx.op('transpose', permutation=perm), [neg_ins])
    
    weight_to_use = mgx_module.add_instruction(migraphx.op('gathernd'), [weight_bcst_t, gathernd_indices])
    x_to_use = mgx_module.add_instruction(migraphx.op('gathernd'), [neg_ins_t, gathernd_indices])
```

**Comment by bpickrel** [comment]
Progress blocked by bug--see issue [149](https://github.com/ROCm/torch_migraphx/issues/149)

**Comment by bpickrel** [comment]
> Progress blocked by bug--see issue [149](https://github.com/ROCm/torch_migraphx/issues/149)

Resolved.

File: py/torch_migraphx/fx/converters/acc_ops_converters.py [hunk 11]
```diff
@@ -112,6 +113,74 @@ def acc_ops_linear(mgx_module, node, args, kwargs):
...
+    # negative of input
+    neg_ins = mgx_module.add_instruction(migraphx.op('neg'), [inp_instr_ref])
+    #
+    # match rank of target to the input: unsqueeze
+    target_ref_unsquoze =  mgx_module.add_instruction(
+        migraphx.op('unsqueeze', axes=list(range(1, ndims))), [target_ref])
+
+    # Prepare to select weight for each target
+    # unsqueeze the weight and broadcast to match input shape 
+    #  TODO: this does not work for ndims > 2 (not currently supported)
```
[the remark this answers is not in the document]
Comment (bpickrel) [reply py/torch_migraphx/fx/converters/acc_ops_converters.py:146]: done

File: tests/fx/converters/test_activations_fx.py [hunk 12]
```diff
@@ -3,6 +3,35 @@
 from fx_test_utils import randbounds, FuncModule, MethodModule, convert_to_mgx, verify_outputs
 
 
+# TODO: test with 3 or more dimensions
+@pytest.mark.parametrize('reduction', [('mean'), ('sum'), ('none')])
+@pytest.mark.parametrize('inp_size, weight_size', [((3, 5), 5), ((3, 5), 0)])
```
[the remark this answers is not in the document]
Comment (bpickrel) [reply tests/fx/converters/test_activations_fx.py:9]: Done (as a separate test).  Note that there's no corresponding dynamo test because the ATen op doesn't allow 1d inputs.

File: py/torch_migraphx/fx/converters/acc_ops_converters.py (hunk 8 above)
Comment (bpickrel) [reply py/torch_migraphx/fx/converters/acc_ops_converters.py]: Let's address this later as out of scope.

File: tests/fx/converters/test_activations_fx.py [hunk 13]
```diff
@@ -3,6 +3,59 @@
...
+    inp = torch.randn(inp_size, dtype=torch.float).cuda()
+    mod = FuncModule(torch.nn.functional.nll_loss, target=target, weight=weight,
+                       reduction = reduction, ignore_index = -100)
+    mgx_mod = convert_to_mgx(mod, [inp])
+    verify_outputs(mod, mgx_mod, [inp])
+
+
+@pytest.mark.parametrize('reduction', [('mean'), ('sum'), ('none')])
+@pytest.mark.parametrize('N, C, weight_size', [(7, 3, 0), (7, 3, 3)])
+def test_nll_loss_1d_fx(N, C, weight_size, reduction):
```
Comment (bpickrel) [inline tests/fx/converters/test_activations_fx.py:41]: Note to @shivadbhavsar :  unfolded 1-d case as a new test.

File: py/torch_migraphx/fx/converters/acc_ops_converters.py [hunk 14]
```diff
@@ -34,6 +34,7 @@
 import migraphx
 import torch
 import numpy as np
+import traceback
```
Comment (shivadbhavsar) [inline py/torch_migraphx/fx/converters/acc_ops_converters.py:37]: remove debug import

File: py/torch_migraphx/fx/converters/utils.py [hunk 15]
```diff
@@ -34,6 +34,7 @@
                      torch_qdtype_to_mgx_enum, torch_dtype_from_mgx,
                      torch_dtype_to_mgx, torch_dtype_to_mgx_enum)
 from ..mgx_module import MGXInstruction
+from numbers import Integral
```
Comment (shivadbhavsar) [inline py/torch_migraphx/fx/converters/utils.py:37]: remove

**Review by shivadbhavsar — COMMENTED** [review]
Looks good, just a few cleanup comments
